README.md
[![DOI](https://zenodo.org/badge/286954083.svg)](https://zenodo.org/badge/latestdoi/286954083)

# ATMODAT Standard Compliance Checker

This is a python library that contains checks to ensure compliance with the ATMODAT Standard.

Its core functionality is based on the [IOOS compliance checker](https://github.com/ioos/compliance-checker). The ATMODAT Standard Compliance Checker library makes use of [cc-yaml](https://github.com/cedadev/cc-yaml), which provides a plugin for the [IOOS compliance checker](https://github.com/ioos/compliance-checker) that generates check suites from YAML descriptions. Furthermore, the [Compliance Check Library](https://github.com/cedadev/compliance-check-lib) is used as the basis to define generic, reusable compliance checks. This repository is an extension of this library as it holds specific checks to ensure compliance with the ATMODAT Standard.

In addition, the compliance to the CF Conventions 1.4 or higher is verified with the [CF checker](https://github.com/cedadev/cf-checker).

<!---
We set up a binder where you can try out the functionalities of the ATMODAT Standard Compliance Checker:

[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/AtMoDat/atmodat_data_checker/HEAD?filepath=notebooks)
-->

## Installation (tested on Linux, macOS and Windows)

1. Install Miniconda or Anaconda (if not already done). Follow the [respective instructions](https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html#regular-installation) for your operating system.


2. Open a Terminal / Anaconda Prompt (under Windows)


3. If not already installed, a version of git is needed to clone repository of the checker from GitHub. It can be installed using conda:
    ```bash
   conda install -c anaconda git
    ```
   Clone this repository:

   ```bash
   git clone https://github.com/AtMoDat/atmodat_data_checker.git
   cd atmodat_data_checker
   ```

4. Create conda environment and install the needed anaconda packages (via mamba):
   ```bash
   conda env create -f environment.yml
   conda activate atmodat
   conda config --add channels conda-forge
   mamba install --file mamba_requirements.txt
   ```

5. Install (and upgrade) atmodat-checker
   ```bash
   pip install -U -e .
   ```

6. Initialize and update the [ATMODAT CVs](https://github.com/AtMoDat/AtMoDat_CVs) submodule, which contains the controlled vocabulary for the ATMODAT Standard.
   ```bash
   git submodule init
   git submodule update  --remote --recursive
   ```

## Update the preinstalled checker 
The ATMODAT Standard Compliance Checker and related packages can be updated to its latest version (GitHub master branch) as follows:
```bash
git checkout master
git fetch || git pull origin master || git submodule update --remote --recursive
mamba update --file mamba_requirements.txt
pip install -U -e .
```

## How to run the ATMODAT Standard Checker

The command `run_checks` can be executed from any directory from within the `atmodat` conda environment. It will perform checks to evaluate the compliance with the ATMODAT Standard.  Compliance with the CF Conventions is checked by a call to the [CF checker](https://github.com/cedadev/cf-checker). The results of the performed checks are provided in the `checker_output` directory. 
By default, `run_checks` assumes writing permissions in the path where the atmodat checker is installed. If this is not the case, you must specify an output directory where you possess writing permissions with the `-op output_path`.

* To check a single file, use:
   ```bash
   run_checks -f file_to_check.nc
   ```
* To run the checker on all `*.nc` files of a directory (including all sub-directories), use:
   ```bash
   run_checks -p file_path
   ```
* To create a summary of the checks performed, add the `-s` flag, e.g.:
   ```bash
   run_checks -s -p file_path
   ```
* To define a custom path into which the checker output shall be written, use:
   ```bash
   run_checks -op output_path -p file_path
   ```
* To define the CF version against which the file(s) shall be checked, use:
   ```bash
   run_checks -cfv 1.6 -p file_path
   ```
   Valid are versions from 1.4 to 1.8. Default is `-cfv auto`.


* To define whether the file(s) shall be checked only against the ATMODAT Standard (AT) or the CF Conventions (CF), specify either `-check AT` or `-check CF`. 
  ```bash
   run_checks -check AT -p file_path
  ```
   Default is `-check both`.


* You can combine different optional arguments, for example:
   ```bash
   run_checks -s -op mychecks -check both -cfv 1.4 -p file_path
  ```
* For more information use:
  ```bash
  run_checks --help
  ```

## Filling global/variable attribute
To ensure compliance with the CF Conventions and the ATMODAT Standard, it might be necessary to adapt the variable and global attributes. To support this process, the `fill_attributes` script is provided as part of the checker which can be used to fill the respective attributes into netCDF files from csv files. Basic examples for these csv files are provided [here](atmodat_checklib/utils/fill_attributes/attribute_templates). These files can be modified to add/change global attributes and to rename variables and add/change their respective attributes. 



### Usage
To run this script, it is necessary to provide the directory where the csv files are stored (`-a csv_directory`). The scirpt can modify a single file or all netCDF files in a given directory. In case you provide a directory, all netCDF files will be identified and modified recursively. Use the `-p` option and give the path/directory of the file(s) to be modified. 

The typical command-line string to run the attribute filler looks like this:
  ```bash
  fill_attributes -a csv_directory -p file_path/directory
  ```
Presently, the original fill will simply be amended to save runtime due to reduced I/O operations. A backup of the original global/variable for each file is created in the `csv_directory`. The original state of the file/directory can be restored using the `-r` option (files will nevertheless not be bit-identical):
  ```bash
  fill_attributes -r -a csv_directory -p file_path/directory
  ```

## Known Issues
* Presently, there is an unresolved issue in the CF checker (v 4.1.0, [see here](https://github.com/cedadev/cf-checker/issues/75)). Until it will get resolved, this issue will output an error that is related to the `formula_terms` attribute in so-called boundary variables. As this is related to the CF checker, we will simply ignore errors that are related to this issue in the `short_summary` output of the checker.

## Contributors

In alphabetic order:

- [Ag Stephens](https://github.com/agstephens)
- [Amandine Kaiser](https://github.com/am-kaiser), [ORCID: 0000-0002-2756-9964](https://orcid.org/0000-0002-2756-9964)
- [Anette Ganske](https://github.com/anganske), [ORCID: 0000-0003-1043-4964](https://orcid.org/0000-0003-1043-4964)
- [Angelika Heil](https://github.com/atmodatcode), [ORCID: 0000-0002-8768-5027](https://orcid.org/0000-0002-8768-5027)
- Angelina Kraft, [ORCID: 0000-0002-6454-335X](https://orcid.org/0000-0002-6454-335X)
- [Daniel Heydebreck](https://github.com/neumannd), [ORCID: 0000-0001-8574-9093](https://orcid.org/0000-0001-8574-9093)
- [Hugo Ricketts](https://github.com/gapintheclouds)
- [Jan Kretzschmar](https://github.com/jkretz), [ORCID: 0000-0002-8013-5831](http://orcid.org/0000-0002-8013-5831)


## Acknowledgements

The ATMODAT Standard Compliance Checker was created within the AtMoDat project (Atmospheric Model Data, <https://www.atmodat.de>). AtMoDat is funded by the German Federal Ministry for Education and Research within the framework of Atmosphären-Modelldaten: Datenqualität, Kurationskriterien und DOI-Branding (FKZ 16QK02A).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
### Added:
### Changed:
### Removed:

## Version [1.3.1] - 2022-06-23
### Added:
- More detailed error message if file suffix ends with `*.NC` (#123)
### Changed:
- Improve `UDUNITS2_XML_PATH` detection in env_util.py (#121)
- Fix CF-Checker related issue and slightly revise short summary output (#126)

## Version [1.3] - 2022-02-17
### Added:
- Remove duplicate from long summary output (#119)
- Add global attribute filler (#102)


## Version [1.2] - 2022-01-28
### Added:
- Check file existence when running on single file (#48)
- Add possibility to run the ATMODAT Standard Compliance Checker in a binder
- Make the AtMoDaT Standard Compliance Checker compatible with Windows (#69)
- Add Changelog.md (#79)
- Add non-recursive file search when path to multiple files is provided (-pnr option) (#81)
- Add information on how to update the ATMODAT Standard Compliance Checker (and related packages)
- Add tests for run_checks.py (check for expected number of missing attributes in demo_data) (#97)
- Add output of version number via -V option (#111)
- Allow for installation form GitHub via pip (#111)
- Output distinct licences in checked files into extra file (#111)

### Changed:
- Add cfunits to environment.yml to fix tests (#52)
- Large revision of README.md (#51)
- Avoid checking for existence the Conventions attribute multiple times (#54)
- Fixing CV checks (#59) and adapt checker to run with new compliance-check-lib version (#68)
- Speed up checking of multiple files (#66)
- Fixing -op option
- Fix command-line string size issue (#80)
- Use mamba to install required conda packages (#83)
- Ignoring errors related to formula_terms in boundary variables (#84)
- Set environment variables during runtime of checker (#85)
- Do not delete content of output directory (#88)
- Restructure the content and use of submodules (#101)
- Make "point" entry in geospatial_lat/lon_resolution a valid entry (#111)
- Revised short summary output (i.e. number of CF checker warnings and reason for failed checks) (#111)
- Fix interpunctation in checker output messages (#111)


## Version [1.1] - 2021-07-07
This is the state before the CHANGELOG.md has been created. From here on, all changes will be tracked.

[Unreleased]: https://github.com/AtMoDat/atmodat_data_checker/compare/v1.3.1...HEAD
[1.3.1]: https://github.com/AtMoDat/atmodat_data_checker/compare/v1.3...v1.3.1
[1.3]: https://github.com/AtMoDat/atmodat_data_checker/compare/v1.2...v1.3
[1.2]: https://github.com/AtMoDat/atmodat_data_checker/compare/v1.1...v1.2
[1.1]: https://github.com/AtMoDat/atmodat_data_checker/compare/v1.0...v1.1


run_checks.py
#!/usr/bin/env python

import argparse
import os
from datetime import datetime
import subprocess
import numpy as np

import atmodat_checklib.utils.output_directory_util as output_directory
import atmodat_checklib.utils.summary_creation_util as summary_creation
from atmodat_checklib.utils.env_util import set_env_variables
from atmodat_checklib import __version__


def main():

    # Set environment variables
    udunits2_xml_path, atmodat_cvs = set_env_variables()
    os.environ['PYESSV_ARCHIVE_HOME'] = os.path.join(atmodat_cvs, 'pyessv-archive')
    os.environ['UDUNITS2_XML_PATH'] = udunits2_xml_path

    idiryml = os.path.join(atmodat_cvs, '')

    # record start time
    start_time = datetime.now()

    # read command line input
    args = command_line_parse()
    verbose = args.verbose
    ifile = args.file
    ipath = args.path
    ipath_norec = args.path_no_recursive
    opath_in = args.opath
    cfversion = args.cfversion
    whatchecks = args.whatchecks
    parsed_summary = args.summary

    # Define output path for checker output
    # user-defined opath
    if opath_in:
        if opath_in.strip() == '/':
            raise RuntimeError('Root directory should not be defined as output path!')
        opath = os.path.abspath(opath_in)
    # predefined opath
    else:
        # default path with subdirectory containing timestamp of check
        opath = os.getcwd()
    opath = os.path.join(opath, 'atmodat_checker_output', '')

    # Define version of CF table against which the files shall be checked.
    # Default is auto --> CF table version parsed from global attribute 'Conventions'.
    if cfversion != 'auto':
        if cfversion not in ('1.4', '1.5', '1.6', '1.7', '1.8'):
            print('User-defined -cfv option invalid; using \'auto\' instead')
            cfversion = 'auto'

    # Let user choose if files shall be checked against (AT) the ATMODAT metadata specifications
    # (excluding the CF checker), (CF) the CF Conventions or (both) against both. Default is both.
    if whatchecks != 'both':
        if whatchecks not in ('AT', 'CF'):
            print('User-defined -check option invalid; using \'both\' instead')
            whatchecks = 'both'

    # Check that either ifile or ipath exist
    if not ifile and not ipath and not ipath_norec:
        raise RuntimeError('No file and path given')

    check_types = ['atmodat', 'CF']
    if whatchecks == 'CF':
        check_types.remove('atmodat')
    elif whatchecks == 'AT':
        check_types.remove('CF')

    # Create directory for checker output
    opath_run = output_directory.create_directories(opath, check_types)

    # Single file
    if ifile and not (ipath or ipath_norec):
        # Check for file ending and add file to list
        files_check = check_file_suffix([ifile])

    # Multiple files
    elif (ipath or ipath_norec) and not ifile:
        # Look for files in given directory
        if ipath:
            files_all = output_directory.return_files_in_directory_tree(ipath)
        else:
            files_all = output_directory.return_files_in_directory(ipath_norec)
        # Check for file ending and add file to list
        files_check = check_file_suffix(files_all)

    # Run global attribute checks
    file_counter = len(files_check)
    run_checks(files_check, verbose, check_types, cfversion, opath_run, idiryml)

    # Create summary of results if specified
    if parsed_summary:
        summary_creation.create_output_summary(file_counter, opath_run, check_types)

    # Create a symbolic link to latest checker output
    latest_link = os.path.join(opath, 'latest')
    if os.path.isdir(latest_link):
        os.unlink(latest_link)
    os.symlink(opath_run, os.path.join(opath, 'latest'))

    # Calculate run time of this script
    print("--- %.4f seconds for checking %s files---" % ((datetime.now() - start_time).total_seconds(), file_counter))


def check_file_suffix(ifiles):
    files_to_check = []
    for file in ifiles:
        if not os.path.isfile(file):
            if not os.path.isdir(file):
                raise RuntimeError(f'File: {file} does not exist')
        else:
            if file.endswith(".nc"):
                files_to_check.append(file)
            elif file.endswith(".NC"):
                raise RuntimeError(f'NetCDF file suffix in {file} must be lower case. Please verify for other files')
    return files_to_check


def utf8len(string_in):
    return len(string_in.encode('utf-8'))


def cmd_string_checker(io_in, idiryml_in):
    # String of input and output files
    ifile_in_string = " ".join(io_in[0])
    ofiles_checker_string = " ".join(io_in[1])

    # Output string creation
    return 'compliance-checker --y ' + idiryml_in + 'atmodat_standard_checks.yml -f json_new -f text ' + \
           ofiles_checker_string + ' --test atmodat_standard:3.0 ' + ifile_in_string


def cmd_string_cf(ifiles_in, cf_version_in):
    ostring = []
    for ifile in ifiles_in:
        ostring.append('cfchecks -v ' + cf_version_in + ' ' + ifile)
    return '; '.join(ostring)


def cmd_string_creation(check_in, ifiles_in, opath_file_in, filenames_base_in, idiryml_in, cf_version_in):
    max_string_len = 131072
    cmd_out = []
    if check_in == 'atmodat':

        # List of output files
        ofiles_checker = ['-o ' + os.path.join(opath_file_in, check_in, filename_base + '_' + check_in + '_result.json')
                          for filename_base in filenames_base_in]

        # Output string creation
        tmp_cmd_out = cmd_string_checker((ifiles_in, ofiles_checker), idiryml_in)
        num_split = int(np.ceil(utf8len(tmp_cmd_out) / max_string_len))
        if num_split > len(ifiles_in):
            num_split = len(ifiles_in)

        # Split if size of string exceeds max_string_len
        if num_split < 2:
            cmd_out.append(tmp_cmd_out)
        else:
            for io_files in zip(np.array_split(ifiles_in, num_split), np.array_split(ofiles_checker, num_split)):
                cmd_out.append(cmd_string_checker(io_files, idiryml_in))

    elif check_in == 'CF':
        tmp_cmd_out = cmd_string_cf(ifiles_in, cf_version_in)
        num_split = int(np.ceil(utf8len(tmp_cmd_out) / max_string_len))
        if num_split > len(ifiles_in):
            num_split = len(ifiles_in)
        if num_split < 2:
            cmd_out.append(tmp_cmd_out)
        else:
            for split_ifiles in np.array_split(ifiles_in, num_split):
                cmd_out.append(cmd_string_cf(split_ifiles, cf_version_in))
    return cmd_out


def run_checks(ifile_in, verbose_in, check_types_in, cfversion_in, opath_file, idiryml_in):
    """run all checks"""
    # Get base filename and output path
    filenames_base = [os.path.basename(os.path.realpath(f)).rstrip('.nc') for f in ifile_in]
    for check in check_types_in:

        # Remove preexisting checker output
        opath_checks = os.path.join(opath_file, check, '')
        for old_file in os.listdir(opath_checks):
            os.remove(os.path.join(opath_checks, old_file))

        if check == 'atmodat':
            cmd_checker = cmd_string_creation(check, ifile_in, opath_file, filenames_base, idiryml_in, cfversion_in)
            for cmd in cmd_checker:
                subprocess.run(cmd, shell=True)
        elif check == 'CF':
            cmd_cf = cmd_string_creation(check, ifile_in, opath_file, filenames_base, idiryml_in, cfversion_in)
            output_string_all = []
            for cmd in cmd_cf:
                output_string_all.append(subprocess.run(cmd, shell=True, capture_output=True, text=True).stdout)
            output_string = ''.join(output_string_all)
            split_string = 'CHECKING NetCDF FILE'
            output_string_files = output_string.split(split_string)[1::]
            for ofile_data_cf in zip(filenames_base, output_string_files):
                ofile_cf = os.path.join(opath_file, 'CF', '') + ofile_data_cf[0] + '_' + check + '_result.txt'
                with open(ofile_cf, 'w', encoding='utf-8') as f_cf:
                    f_cf.write(split_string + ofile_data_cf[1])
    for filename_base in filenames_base:
        for check in check_types_in:
            file_verbose = os.path.join(opath_file, check, '') + filename_base + '_' + check + '_result.txt'
            if verbose_in:
                with open(file_verbose, 'r', encoding='utf-8') as f_verbose:
                    if check == 'CF':
                        print('')
                        print('==============================================================================')
                        print('===============================CF-Checker=====================================')
                        print('==============================================================================')
                        print('')
                        print(f_verbose.read())
                    elif check == 'atmodat':
                        print('')
                        print('==============================================================================')
                        print('===============================AtMoDat-Checks=================================')
                        print('==============================================================================')
                        print('')
                        print(f_verbose.read())
            # Clean-up
            if check == 'atmodat':
                if os.path.isfile(file_verbose):
                    os.remove(file_verbose)
    return


def command_line_parse():
    """parse command line input"""
    # Parser for command line options
    parser = argparse.ArgumentParser(description="Run the AtMoDat checks suits.")
    parser.add_argument("-v", "--verbose",
                        help="Print output of checkers (longer runtime due to double call of checkers)",
                        action="store_true",
                        default=False)
    parser.add_argument("-op", "--opath", help="Define custom path where checker output shall be written",
                        default=False)
    parser.add_argument("-cfv", "--cfversion", help="Define custom CF table version against which the file shall be "
                                                    "checked. Valid are versions from 1.3 to 1.8.  "
                                                    "Example: \"-cfv 1.6\". Default is 'auto'",
                        default='auto')
    parser.add_argument("-check", "--whatchecks", help="Define if AtMoDat or CF check or both shall be executed. "
                                                       "Valid options: AT, CF, both. Example: \"-check CF\". "
                                                       "Default is 'both'",
                        default='both')
    parser.add_argument("-s", "--summary", help="Create summary of checker output",
                        action="store_true",
                        default=False)
    parser.add_argument('-V', '--version', action='version',
                        version=f'ATMODAT Standard Compliance Checker Version: {__version__}')
    group = parser.add_mutually_exclusive_group()
    group.add_argument("-f", "--file", help="Processes the given file")
    group.add_argument("-p", "--path", help="Processes all files in a given path and subdirectories "
                                            "(recursive file search)")
    group.add_argument("-pnr", "--path_no_recursive", help="Processes all files in a given directory")

    return parser.parse_args()


if __name__ == "__main__":
    main()
